x/ecocredit/server/proposal.go

```diff
@@ -0,0 +1,45 @@
+package server
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	govtypes "github.com/cosmos/cosmos-sdk/x/gov/types"
+
+	"github.com/regen-network/regen-ledger/x/ecocredit/core"
+	"github.com/regen-network/regen-ledger/x/ecocredit/marketplace"
+)
+
+// ProposalKeeper defines methods for ecocredit gov handlers.
+type ProposalKeeper interface {
+	AddCreditType(ctx sdk.Context, ctp *core.CreditTypeProposal) error
+	AddAskDenom(ctx sdk.Context, proposal *marketplace.AllowedDenomProposal) error
+}
+
+func (s serverImpl) AddAskDenom(ctx sdk.Context, proposal *marketplace.AllowedDenomProposal) error {
+	return s.marketplaceKeeper.AddAllowedDenom(ctx, proposal)
+}
+
+func (s serverImpl) AddCreditType(ctx sdk.Context, ctp *core.CreditTypeProposal) error {
+	return s.coreKeeper.AddCreditType(ctx, ctp)
+}
+
+func NewProposalHandler(k ProposalKeeper) govtypes.Handler {
+	return func(ctx sdk.Context, content govtypes.Content) error {
+		switch c := content.(type) {
+		case *core.CreditTypeProposal:
+			return handleCreditTypeProposal(ctx, k, c)
+		case *marketplace.AllowedDenomProposal:
+			return handleAllowedDenomProposal(ctx, k, c)
+		default:
+			return sdkerrors.ErrUnknownRequest.Wrapf("unrecognized proposal content type: %T", c)
+		}
+	}
+}
+
+func handleAllowedDenomProposal(ctx sdk.Context, k ProposalKeeper, proposal *marketplace.AllowedDenomProposal) error {
+	return k.AddAskDenom(ctx, proposal)
+}
+
+func handleCreditTypeProposal(ctx sdk.Context, k ProposalKeeper, proposal *core.CreditTypeProposal) error {
+	return k.AddCreditType(ctx, proposal)
+}
```

Comment:
Same as above and for both proposals. The proposals should probably describe what action is taking place, i.e. adding an allowed denom and adding a credit type.

Reply:
i was following the pattern set [here](https://github.com/cosmos/cosmos-sdk/blob/6f070623741fe0d6851d79ada41e6e2b1c67e236/x/upgrade/handler.go#L28-L36) in cosmos-sdk and these: [1](https://github.com/osmosis-labs/osmosis/blob/bdff1dacaabe8055f52fc40a812c92df5a6c9547/x/pool-incentives/handler.go), [2](https://github.com/osmosis-labs/osmosis/blob/ccccc53426483188d0bc590e570e181a978bb51c/x/superfluid/proposal_handler.go), [3](https://github.com/osmosis-labs/osmosis/blob/bdff1dacaabe8055f52fc40a812c92df5a6c9547/x/txfees/handler.go) from osmosis. 

seems they all just do `handle` +proposal message name. I'm okay with deviating from this, but thought it might be good to consider these

Reply:
The examples in the sdk describe the action, i.e. "upgrade" and "cancel". Same with the ones in Osmosis, i.e. "replace", "update", "set", "remove", and "update". It would not be deviating but rather aligning if we use "add".

Reply:
Or if we use "allow" and "add".

Reply:
The message name is now `AllowDenomProposal` so I think that resolves it. See https://github.com/regen-network/regen-ledger/pull/1072#discussion_r866109614.